### ISSUE TITLE
Implement support for vendoring procedures in the assembler

### ISSUE THREAD

**Issue by bobbinth** [0xPolygonMiden/miden-vm#1435]
Currently, when the assembler assembles libraries/programs, it assumes that all external libraries will be available to the processor to execute the assembled code. However, this may not be true in all instances.

So, when adding external libraries to the assembler, we need to specify whether a given library is expected to be available to the processor or should be vendored (i.e., copied into the assembled program/library).

**Comment by bitwalker** [comment]
This is fairly straightforward to implement, we'll need another API for adding vendored libraries, and for those libraries we'll want to take ownership over the `CompiledLibrary` (or clone it), we'd track these separately from non-vendored libraries. Then, when we encounter a reference to a procedure whose definition is in a vendored library, we do a depth-first, post-order traversal of the MastForest of that library, from the procedure root, copying all of the nodes (after all their children are processed) into the `MastForest` being built, rewriting any `MastNodeId`s in the process. This will ensure that we only vendor what is strictly needed. We can of course skip the traversal for any node that is already in the forest, so long as that node is not `External` (unless the `External` node is not present in any of the vendored libraries).

**Comment by bobbinth** [comment]
I think one challenge we'd need to address about vendored libraries is that it may be expensive to proactively clone all MAST forests (we kind of do this right now for non-vendored libraries, but I see this more like a temporary solution). I think building a set of `ModuleInfo` objects for all modules that may potentially be relevant to the program/library that is being assembled is probably cheap enough because `ModuleInfo` struct is pretty light-weight. But we'd probably want to avoid doing the same for all potentially relevant MAST forests.

I think we can address this by introducing something like `AssemblyContext` I mentioned in https://github.com/0xPolygonMiden/miden-vm/issues/1433#issuecomment-2274303514. Here, the assembler would never actually take ownership of libraries, but would use the context to build `ModuleInfo` structs and extract MASTs only for relevant procedures.

**Comment by bitwalker** [comment]
I suspect we could easily refactor the API to take `CompiledLibrary` by reference (and in general, it probably makes sense to always wrap them in `Rc` or `Arc` anyway, as they are immutable, and that would vastly simplify this issue), since almost by definition they would outlive the `Assembler` due to its one-shot nature.

The `ModuleGraph` itself doesn't need ownership of them, only a reference, the main pain point to using plain references versus smart pointers is the need to refactor a bunch of APIs to have explicit lifetimes. So my first instinct would be to evaluate if we can just require the use of reference-counted `CompiledLibrary` references.

**Comment by bobbinth** [comment]
I actually didn't realize that `add_modules()` already allows for vendoring code - nice to know we have this option! This issue though, was more about vendoring already compiled libraries. I can imagine a number of cases when we'd want to distribute compiled libraries rather than the original source code (this could be relevant for both libraries generated from hand-written MASM as well as libraries generated by the compiler).

The way I was imagining it was by modifying the existing `add_library()` to look something like this:
```rust
pub fn add_library(&mut self, library: impl AsRef<Library>, is_vendored: bool) -> Result<(), Report> {
    ...
}
```
But we can also add a new method like so:
```rust
pub fn add_vendored_library(&mut self, library: impl AsRef<Library>) -> Result<(), Report> {
    ...
}
```
The second option may actually be better as it doesn't introduce a breaking change and seems to be cleaner overall.

Re internal implementation, I was originally imagining that we'd copy the relevant procedures into the generated `MastForest` during assembly. But if we can use `MastForest::merge()` and get similar result, that would work as well.

I think the important thing is to make sure we don't end up vendoring procedures which are not relevant. For example, if `A` exports procedures `x`, `y`, `z`, and `B` uses `A` as a dependency but references only procedure `x`, then when we build a `.masl` file for `B`, it should not contain the code for procedures `y` and `z`.

Another thing to consider is how to vendor advice map data that may be present in a vendored dependency. I think for now we can just always copy this data into the assembled library/program. In the future, we could somehow mark specific procedures which rely on the advice map data and only copy the relevant parts of the advice map during assembly.

**Comment by paracetamolo** [comment]
It could be the case that my library `B` adds more functionalities to library `A`, so I want to export all the procedures of `A` and add the ones in `B`. 
Is the convention that library `B` should re-export the procedures of `A`?

**Comment by bobbinth** [comment]
> It could be the case that my library `B` adds more functionalities to library `A`, so I want to export all the procedures of `A` and add the ones in `B`.
> Is the convention that library `B` should re-export the procedures of `A`?

Yeah, in this case we'd have to manually re-export all procedures. Definitely not ideal, and maybe in the future we'd add syntax for re-exporting entire modules somehow.

I do think that the use case of re-exporting entire libraries but with some additional functionality would probably be less common then using just a small subset from multiple dependencies in downstream libraries/programs.

**Comment by paracetamolo** [comment]
After a bit more investigating, the problem is that the only persistent structure in the assembler is the module graph which only takes "source" modules, so there is no easy way now to store a compiled library. I see two solutions.

1. Modify the `assemble_program` function to take a list of libraries to vendor.  The reason is that when we assemble the program it's the only time when a mast forest is created and so it's the only time when it can be merged with the forest of the compiled libraries. This would break the the current API and I'm not sure it's the right API to the future.
I implemented this quickly and it's a very short diff, need to test it more though.

2. We could extend the `WrappedModule` used in `ModuleGraph` to take also a library. In this way we could store vendored libraries just as we store ast modules. It would not impact the external API. I tried this and it requires much more internal work, at least we need to also change `WrappedProcedure` and I'm not sure what else.

**Comment by bobbinth** [comment]
I was thinking it would be more like the second approach. Though, we can always pretty easily transform the first approach into the second. For example, we could keep track of the libraries to be vendored separately (e.g., in a separate field of the `ModuleGraph`) and then on `assemble_program()`, `assemble_library()`, `assemble_kernel()`, we'd get this libraries and treat them as if they were passed in via arguments.

However, a bigger question for me is whether merging actually works as we want it to. Specifically, if we merge forest `A` and forest `B`, wouldn't the resulting forest be a union of procedures in `A` and `B`? Assuming `A` depends on `B`, how would we ensure that we only retain the relevant parts of `B` and drop everything else?

**Comment by paracetamolo** [comment]
You are right that merging two libraries is just the union.

If we are assembling a library A and we want to vendor into it library B, we won't get the dropping of unnecessary parts that we want.

If we are assembling a program, then the function `compile_subgraph` will take care of picking from the vendored libraries only the dependencies strictly needed by the program.

**Comment by bobbinth** [comment]
> If we are assembling a library A and we want to vendor into it library B, we won't get the dropping of unnecessary parts that we want.

Yeah, makes sense. So, we need to come up with a solution which would only retain the relevant parts of `B` when vendoring it for `A` (this would be a common use case for building account components).

**Comment by paracetamolo** [comment]
I gave it a go here https://github.com/0xPolygonMiden/miden-vm/pull/1643
Like discussed above, the user can use `vendor_library` to append compiled libraries to the assembler. Once `assemble_{program,library}` is called, the resulting MAST is merged with the vendored libraries and then there is a pass to remove unnecessary roots and all their descendant nodes.

**Comment by bobbinth** [comment]
Closed by #1643.
